hub/util/version_control.py

```diff
@@ -0,0 +1,216 @@
+        "commit_id"
+    ]
+    version_state["commit_node_map"][version_state["commit_id"]] = new_node
+    save_version_info(version_state, storage)
+    copy_metas(
+        stored_commit_id, version_state["commit_id"], storage, version_state["tensors"]
+    )
+    storage.flush()
+    load_meta(storage, version_state)
+
+
+def checkout(
+    version_state: Dict[str, Any],
+    storage: LRUCache,
+    address: str,
+    create: bool = False,
+) -> None:
+    """Modifies the version state to reflect the checkout and also copies required data to the new branch directory if a new one is being created."""
+    if address in version_state["branch_commit_map"].keys():
+        if create:
+            raise CheckoutError(f"Can't create new branch, '{address}' already exists.")
+        version_state["branch"] = address
+        version_state["commit_id"] = version_state["branch_commit_map"][address]
+        version_state["commit_node"] = version_state["commit_node_map"][
+            version_state["commit_id"]
+        ]
+    elif address in version_state["commit_node_map"].keys():
+        if create:
+            raise CheckoutError(
+                f"Can't create new branch, commit '{address}' already exists."
+            )
+        version_state["commit_id"] = address
+        version_state["commit_node"] = version_state["commit_node_map"][address]
+        version_state["branch"] = version_state["commit_node"].branch
+    elif create:
+        # if the original commit is head of the branch and has data, auto commit and checkout to original commit before creating new branch
+        auto_commit(version_state, storage, address)
+        original_commit_id = version_state["commit_id"]
+        new_commit_id = generate_hash()
+        new_node = CommitNode(address, new_commit_id)
+        new_node.parent = version_state["commit_node"]
+        version_state["commit_id"] = new_commit_id
+        version_state["commit_node"] = new_node
+        version_state["branch"] = address
+        version_state["commit_node_map"][new_commit_id] = new_node
+        version_state["branch_commit_map"][address] = new_commit_id
+        save_version_info(version_state, storage)
+        copy_metas(
+            original_commit_id,
+            new_commit_id,
+            storage,
+            version_state["tensors"],
+        )
+        storage.flush()
+    else:
+        raise CheckoutError(
+            f"Address {address} not found. If you want to create a new branch, use checkout with create=True"
+        )
+    load_meta(storage, version_state)
+
+
+def copy_metas(
+    src_commit_id: str, dest_commit_id: str, storage: LRUCache, tensors: Dict
+) -> None:
+    """Copies meta data from one commit to another."""
+    src_dataset_meta_key = get_dataset_meta_key(src_commit_id)
+    dest_dataset_meta_key = get_dataset_meta_key(dest_commit_id)
+    storage[dest_dataset_meta_key] = storage[src_dataset_meta_key].copy()
```

Comment:
Is .copy() here necessary?

Reply:
Yup, copy is needed otherwise both metas use the same object and modifying any one of these leads to both being modified, learnt that the hard way